### PR TITLE
Report rude edit when editing async/iterator method and SM attribute is not available

### PR DESCRIPTION
**Pull request by tmat**
Adds logic to IDE EnC analyzer to report a rude edit. 
Makes the compiler more robust in case the attributes are not well-formed or missing for other reasons.

Fixes https://github.com/dotnet/roslyn/issues/9119

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-interactive @dotnet/roslyn-compiler Please review.

**Comment by tmat** [comment]
@dotnet-bot retest prtest/win/dbg/eta please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_dbg_eta/4345/
// Retest reason: nuget restore failed

**Comment by MattGertz** [comment]
Approved pending code reviews.

File: src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb [hunk 1]
```diff
@@ -15,6 +15,26 @@ Imports Xunit
...
+                                      Optional references As IEnumerable(Of MetadataReference) = Nothing,
+                                      Optional options As VisualBasicCompilationOptions = Nothing,
+                                      Optional assemblyName As String = Nothing) As VisualBasicCompilation
+        If options Is Nothing Then
+            options = TestOptions.ReleaseDll
+        End If
+
+        ' Using single-threaded build if debugger attached, to simplify debugging.
+        If Debugger.IsAttached Then
+            options = options.WithConcurrentBuild(False)
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb]: This can actually affect tests that specifically test Concurrent build.

Comment (tmat) [inline src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb]: This is a copy from C#.

File: src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb [hunk 2]
```diff
@@ -15,6 +15,26 @@ Imports Xunit
 
 Friend Module CompilationUtils
 
+    Public Function CreateCompilation(trees As IEnumerable(Of SyntaxTree),
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb]: If this is a helper with a very special behavior, the name should reflect that and the comment should clearly explain what is special about it.

File: src/Compilers/Core/Portable/Emit/EditAndContinue/EncVariableSlotAllocator.cs [hunk 3]
```diff
@@ -205,9 +224,18 @@ public override bool TryGetPreviousHoistedLocalSlotIndex(SyntaxNode currentDecla
...
-        public override bool TryGetPreviousAwaiterSlotIndex(Cci.ITypeReference currentType, out int slotIndex)
+        public override bool TryGetPreviousAwaiterSlotIndex(Cci.ITypeReference currentType, DiagnosticBag diagnostics, out int slotIndex)
         {
-            Debug.Assert(_awaiterMapOpt != null);
+            // Well-formed state machine attribute wasn't found in the baseline (the type is missing or bad).
+            // Should rarely happen since the IDE reports a rude edit if the attribute type doesn't exist.
+            if (_awaiterMapOpt == null)
+            {
+                // TODO: better error message https://github.com/dotnet/roslyn/issues/9196
+                diagnostics.Add(_messageProvider.CreateDiagnostic(_messageProvider.ERR_ModuleEmitFailure, NoLocation.Singleton, _previousTopLevelMethod.ContainingModule.Name));
```
Comment (cston) [inline src/Compilers/Core/Portable/Emit/EditAndContinue/EncVariableSlotAllocator.cs]: Consider extracting a method.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 4]
```diff
@@ -3687,6 +3700,37 @@ private bool AreEquivalentClosureScopes(SyntaxNode oldScopeOpt, SyntaxNode newSc
...
+            List<RudeEditDiagnostic> diagnostics)
+        {
+            if (!updatedInfo.OldHasStateMachineSuspensionPoint)
+            {
+                return;
+            }
+
+            // only methods and anonymous functions may be async/iterators machines:
+            var stateMachineAttributeQualifiedName = ((IMethodSymbol)oldMember).IsAsync ?
+                "System.Runtime.CompilerServices.AsyncStateMachineAttribute" :
```
Comment (AlekseyTs) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Use AttributeDescription.AsyncStateMachineAttribute to get this name. Same for IteratorStateMachineAttribute

Comment (AlekseyTs) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Assuming it is accessible.

Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: It's not. This is IDE layer.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 5]
```diff
@@ -3687,6 +3700,37 @@ private bool AreEquivalentClosureScopes(SyntaxNode oldScopeOpt, SyntaxNode newSc
...
+                "System.Runtime.CompilerServices.IteratorStateMachineAttribute";
+
+            // We assume that the attributes, if exist, are well formed.
+            // If not an error will be reported during emit.
+            if (oldMember.ContainingAssembly.GetTypeByMetadataName(stateMachineAttributeQualifiedName) == null)
```
Comment (AlekseyTs) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: This helper doesn't look for a type the same way compiler looks for attributes. Also, these attributes are optional, so no error will be reported by compiler if they are not well formed.

Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: That is expected. See comment above.

File: src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb]: This doesn't make things right.

Comment (tmat) [inline src/Compilers/Test/Utilities/VisualBasic/CompilationTestUtils.vb]: It doesn't make them wrong either. I can follow up cleaning up both helpers. For now I'm gonna keep it in sync.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: The comment is claiming that an error will be reported during emit. I do not believe compiler reports errors for these attributes.

Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: The EnC logic will: https://github.com/dotnet/roslyn/pull/9168/files#diff-e8dfca4da0b5febea0749842089c9424R195

**Comment by cston** [comment]
LGTM

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Please clarify the comment by making it clear what layer is responsible for reporting the error.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 5 above)
Comment (tmat) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: OK.

**Pull request merged**